testcases/kernel/syscalls/madvise/madvise08.c

```diff
@@ -0,0 +1,255 @@
+	int dfd;
+};
+
+static struct test_case tc = {
+	.fd = 0,
+	.fmem = NULL,
+	.dfd = 0,
+	.ycount = 1280,
+	.cpattern = NULL,
+};
+
+static void setup(void)
+{
+	const char *dumpf = "dump-%p";
+	char cwd[1024];
+	char tmpcpattern[1048];
+	char *fmemc;
+	int i;
+	unsigned int filter;
+	struct rlimit limit;
+
+	tc.fmemsize = tc.ycount + 2;
+
+	SAFE_GETRLIMIT(RLIMIT_CORE, &limit);
+	if (limit.rlim_cur == 0) {
+		tst_brk(TCONF, "Core dump size resource limit is zero.");
+	} else if (limit.rlim_cur < 100000){
+		tst_res(TCONF, "Core dump size limit may be too low.");
+	}
+
+	switch (prctl(PR_GET_DUMPABLE)) {
+	case 0:
+		tst_brk(TCONF, "Process is not dumpable.");
+	case 1:
+		break;
+	default:
+		tst_brk(TBROK | TERRNO, "prctl(PR_GET_DUMPABLE)");
+	}
+
+	SAFE_FILE_SCANF(CORE_FILTER, "%x", &filter);
+	if (!(0x1 & filter))
+		tst_brk(TCONF, "Anonymous private memory is not dumpable.");
+
+	SAFE_FILE_SCANF(CORE_PATTERN, "%m[^\n]", &tc.cpattern);
+	tst_res(TINFO, "System core pattern is '%s'", tc.cpattern);
+
+	SAFE_GETCWD(cwd, sizeof(cwd));
+	snprintf(tmpcpattern, sizeof(tmpcpattern), "%s/%s", cwd, dumpf);
+	tst_res(TINFO, "Temporary core pattern is '%s'", tmpcpattern);
+	SAFE_FILE_PRINTF(CORE_PATTERN, "%s", tmpcpattern);
+
+	tc.fmem = SAFE_MMAP(NULL,
+			    tc.fmemsize,
+			    PROT_READ | PROT_WRITE,
+			    MAP_ANONYMOUS | MAP_PRIVATE,
+			    -1,
+			    0);
+
+	/*Write a generated character sequence to the mapped memory,
+	  which we later look for in the core dump.*/
+	fmemc = (char *)tc.fmem;
+	*fmemc = 'x';
+	for (i = 0; i < tc.ycount; i++)
+		fmemc[i + 1] = 'y';
+	fmemc[++i] = 'z';
+}
+
+static void cleanup(void)
+{
+	if (tc.cpattern != NULL) {
+		if(FILE_PRINTF(CORE_PATTERN, "%s", tc.cpattern)) {
+			tst_brk(TBROK,
+				"COULD NOT RESTORE "CORE_PATTERN"! It used to be '%s'",
+				tc.cpattern);
+		}
+		free(tc.cpattern);
+	}
+
+	if (tc.fmem != NULL && munmap(tc.fmem, tc.fmemsize))
+		tst_res(TBROK | TERRNO, "munmap(tc.fmem)");
+	tc.fmem = NULL;
+
+	if (tc.dfd > 0 && close(tc.dfd))
+		tst_res(TBROK | TERRNO, "close(tc.dfd)");
+}
+
+static int find_sequence(int pid)
+{
+	char expectc = 'x';
+	ssize_t read, pos = 0;
+	char rbuf[1024];
+	int ycount = 0;
+	char dumpname[256];
+
+	snprintf(dumpname, 256, "dump-%d", pid);
+	tst_res(TINFO, "Dump file should be %s", dumpname);
+	if (access(dumpname, F_OK))
+		tst_brk(TBROK | TERRNO, "Dump file was not found.");
+
+	tc.dfd = SAFE_OPEN(dumpname, O_RDONLY);
+
+	read = SAFE_READ(0, tc.dfd, &rbuf, sizeof(rbuf));
+	while (read) {
+		switch (rbuf[pos]) {
+		case 'x':
+			ycount = 0;
+			expectc = 'y';
+			break;
+		case 'y':
+			if (expectc == 'y') {
+				ycount++;
+			} else {
+				expectc = 'x';
+				break;
+			}
+
+			if (ycount == tc.ycount)
+				expectc = 'z';
+			break;
+		case 'z':
+			if (expectc == 'z')
+				return 1;
+		default:
+			expectc = 'x';
+		}
+		if (++pos >= read) {
+			read = SAFE_READ(0, tc.dfd, &rbuf, sizeof(rbuf));
+			pos = 0;
+		}
+	}
+
+	SAFE_CLOSE(tc.dfd);
+	tc.dfd = 0;
```

Comment:
The `SAFE_CLOSE()` sets the file descriptor to -1 already. No need to reset it yourself.